Runtime/Scripts/DataCollection/RGDataCollection.cs

```diff
@@ -0,0 +1,135 @@
+        {
+            // Name the session, and setup a temporary directory for all data
+            _sessionName = Guid.NewGuid().ToString();
+            _logDataPoints = new ConcurrentQueue<LogDataPoint>();
+            _replayDataForTick = new ();
+            _parent = parent;
+
+            // Setup all listener-based data collection (as opposed to tick-based)
+            Application.logMessageReceivedThreaded += CaptureLog;
+
+            Debug.Log("RGSession created and started!");
+        }
+
+        public void CaptureScreenshot(long tick)
+        {
+            Debug.Log($"Captured screenshot at tick {tick}");
+            string path = GetSessionDirectory($"screenshots/{tick}.jpg");
+
+            var texture = ScreenCapture.CaptureScreenshotAsTexture(1);
+
+            // Encode the texture into a jpg byte array
+            byte[] bytes = texture.EncodeToJPG(100);
+
+            // Save the byte array as a jpg file
+            File.WriteAllBytes(path, bytes);
+
+            // Destroy the texture to free up memory
+            Object.Destroy(texture);
+        }
+
+        public void CaptureReplayDataPoint(RGTickInfoData tickInfo, long playerId, RGActionRequest[] actions, RGValidationResult[] validations)
+        {
+            _replayDataForTick[tickInfo.tick] = new ReplayDataPoint(tickInfo, playerId, actions, validations);
+        }
+
+        public void CaptureLog(string logString, string stackTrace, LogType type)
+        {
+            _logDataPoints.Enqueue(new LogDataPoint(logString, stackTrace, type));
+        }
+
+        public void RecordSession(long botInstanceId, RGClientConnectionType rgClientConnectionType)
+        {
+            Debug.Log("Ending RGSession...");
+            
+            Debug.Log("Saving all replay data files and zipping...");
+            foreach (var replayDataPoint in _replayDataForTick)
+            {
+                var replayData = JsonConvert.SerializeObject(replayDataPoint.Value, Formatting.None, new JsonSerializerSettings
+                {
+                    ReferenceLoopHandling = ReferenceLoopHandling.Ignore
+                });
+                File.WriteAllText(GetSessionDirectory($"replay_data/{replayDataPoint.Key}.txt"), replayData);
+            }
+            // Zip the replay data text files
+            
+
+            // Now, begin uploading
+            Debug.Log($"Uploading data collected for bot {botInstanceId}...");
+            
+            // First, upload all of the screenshots
+            var screenshotFiles = Directory.GetFiles(GetSessionDirectory($"screenshots"));
+            foreach (var screenshotFilePath in screenshotFiles)
+            {
+                var screenshotFile = Path.GetFileName(screenshotFilePath);
+                var tick = long.Parse(screenshotFile.Split(".")[0]);
+                RGServiceManager.GetInstance()?.UploadScreenshot(
+                    botInstanceId, tick, screenshotFilePath,
```

Comment:
1. There's no guarantee that `TcpClient.Close` has finished closing the connection when it returns, so `RGClientConnection_Remote.Close` also can't provide that guarantee.
2. `UploadScreenshot` triggers a `UnityWebRequest`, which is an async operation but doesn't `await` it. So it will return from this method before the upload has completed.

Async operations **that actually use a background thread** like `UnityWebRequest` work by kicking off the request in the background and polling for status on each frame. If you `await` the `Task` returned by `UploadScreenshot`, execution of the rest of this method is deferred until the next frame _after the request completes_

Reply:
In addition to @analogrelay's analysis....

The core issue you have @vontell is with where you are calling `RecordSession` in `RGBotServerListener.EndClientConnection`

For `REMOTE` bots only, you need to call `RecordSession` in the error/success handler of the call to `RGService` to end the client connection. <- This call is what creates the bot history record.

Something akin to this...
```csharp
// --- Don't call this for local bots ...
if (clientConnection.Type == RGClientConnectionType.REMOTE)
{
    _ = RGServiceManager.GetInstance()?.StopBotInstance(
        clientId,
        () => {
            // Upload the replay data for this bot
            _dataCollection.RecordSession(clientId, clientConnection.Type);
        },
        () => {
            // Upload the replay data for this bot
            _dataCollection.RecordSession(clientId, clientConnection.Type);
        }
    );
} else 
{
    // Upload the replay data for this bot
    _dataCollection.RecordSession(clientId, clientConnection.Type);
}
```